File: dev-server.js
#!/usr/bin/env node
/**
 * NgKore Website - Clean URL Development Server
 * This server serves clean URLs without .html extensions appearing in the browser
 */

const http = require("http");
const fs = require("fs");
const path = require("path");
const url = require("url");

// MIME types mapping
const mimeTypes = {
  ".html": "text/html",
  ".js": "text/javascript",
  ".css": "text/css",
  ".json": "application/json",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".svg": "image/svg+xml",
  ".ico": "image/x-icon",
  ".pdf": "application/pdf",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
  ".ttf": "font/ttf",
  ".eot": "application/vnd.ms-fontobject",
  ".webmanifest": "application/manifest+json",
};

function getMimeType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  return mimeTypes[ext] || "application/octet-stream";
}

function serveFile(res, filePath) {
  fs.readFile(filePath, (err, data) => {
    if (err) {
      console.error(`Error reading file ${filePath}:`, err);
      serve404(res);
      return;
    }

    const mimeType = getMimeType(filePath);

    res.writeHead(200, {
      "Content-Type": mimeType,
      "Content-Length": data.length,
      "X-Content-Type-Options": "nosniff",
      "X-Frame-Options": "DENY",
      "X-XSS-Protection": "1; mode=block",
    });

    res.end(data);
  });
}

function serve404(res) {
  const html = `
    <!DOCTYPE html>
    <html>
    <head>
        <title>404 - Page Not Found</title>
        <style>
            body { font-family: Arial, sans-serif; margin: 50px; }
            h1 { color: #4A90E2; }
        </style>
    </head>
    <body>
        <h1>404 - Page Not Found</h1>
        <p>The requested page could not be found.</p>
        <p><a href="/">Return to Home</a></p>
    </body>
    </html>
    `;

  res.writeHead(404, {
    "Content-Type": "text/html",
    "Content-Length": html.length,
  });
  res.end(html);
}

function createServer(port) {
  const server = http.createServer((req, res) => {
    const parsedUrl = url.parse(req.url);
    let pathname = decodeURIComponent(parsedUrl.pathname);

    // Remove leading slash
    if (pathname.startsWith("/")) {
      pathname = pathname.substring(1);
    }

    // Handle root path
    if (!pathname || pathname === "index") {
      pathname = "index.html";
    }

    const filePath = path.join(__dirname, pathname);

    // Check if the exact file exists
    fs.access(filePath, fs.constants.F_OK, (err) => {
      if (!err) {
        // File exists, serve it
        serveFile(res, filePath);
        return;
      }

      // If file doesn't exist and has no extension, try adding .html
      if (!path.extname(pathname)) {
        const htmlPath = filePath + ".html";
        fs.access(htmlPath, fs.constants.F_OK, (htmlErr) => {
          if (!htmlErr) {
            serveFile(res, htmlPath);
          } else {
            serve404(res);
          }
        });
      } else {
        serve404(res);
      }
    });

    // Log the request
    console.log(`[${new Date().toISOString()}] ${req.method} ${req.url}`);
  });

  server.listen(port, () => {
    console.log("NgKore Website - Clean URL Server");
    console.log(`Serving from: ${__dirname}`);
    console.log(`Server URL: http://localhost:${port}`);
    console.log("Clean URLs enabled (no .html extensions in browser)");
    console.log("\nTest these clean URLs:");
    console.log(`   http://localhost:${port}/`);
    console.log(`   http://localhost:${port}/pages/about`);
    console.log(`   http://localhost:${port}/pages/contact`);
    console.log(`   http://localhost:${port}/pages/expertise`);
    console.log(
      `   http://localhost:${port}/pages/expertise/telecom-architecture`
    );
    console.log("\n Press Ctrl+C to stop the server\n");
  });

  return server;
}

// Get port from command line or use default
const port = process.argv[2] ? parseInt(process.argv[2]) : 8000;

if (isNaN(port)) {
  console.error("Invalid port number. Using default port 8000.");
  createServer(8000);
} else {
  createServer(port);
}

File: dev-server.py
#!/usr/bin/env python3
"""
NgKore Website - Clean URL Development Server
This server serves clean URLs without .html extensions appearing in the browser
"""

import http.server
import socketserver
import os
import sys
import mimetypes
from urllib.parse import urlparse, unquote


class CleanURLHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        # Parse the requested path
        parsed_url = urlparse(self.path)
        requested_path = unquote(parsed_url.path)

        # Remove leading slash
        if requested_path.startswith('/'):
            requested_path = requested_path[1:]

        # Handle root path
        if not requested_path or requested_path == 'index':
            file_path = 'index.html'
        else:
            # Check if the requested path exists as a file
            if os.path.isfile(requested_path):
                file_path = requested_path
            # If not, try adding .html extension
            elif os.path.isfile(requested_path + '.html'):
                file_path = requested_path + '.html'
            else:
                self.send_404()
                return

        # Serve the file
        self.serve_file(file_path)

    def serve_file(self, file_path):
        try:
            with open(file_path, 'rb') as f:
                content = f.read()

            # Get MIME type
            mime_type, _ = mimetypes.guess_type(file_path)
            if mime_type is None:
                mime_type = 'application/octet-stream'

            # Send response
            self.send_response(200)
            self.send_header('Content-Type', mime_type)
            self.send_header('Content-Length', str(len(content)))

            # Security headers
            self.send_header('X-Content-Type-Options', 'nosniff')
            self.send_header('X-Frame-Options', 'DENY')
            self.send_header('X-XSS-Protection', '1; mode=block')

            self.end_headers()
            self.wfile.write(content)

        except FileNotFoundError:
            self.send_404()
        except Exception as e:
            self.send_500(str(e))

    def send_404(self):
        self.send_response(404)
        self.send_header('Content-Type', 'text/html')
        self.end_headers()
        html = '''
        <!DOCTYPE html>
        <html>
        <head>
            <title>404 - Page Not Found</title>
            <style>
                body { font-family: Arial, sans-serif; margin: 50px; }
                h1 { color: #4A90E2; }
            </style>
        </head>
        <body>
            <h1>404 - Page Not Found</h1>
            <p>The requested page could not be found.</p>
            <p><a href="/">Return to Home</a></p>
        </body>
        </html>
        '''
        self.wfile.write(html.encode())

    def send_500(self, error):
        self.send_response(500)
        self.send_header('Content-Type', 'text/html')
        self.end_headers()
        html = f'''
        <!DOCTYPE html>
        <html>
        <head>
            <title>500 - Server Error</title>
            <style>
                body {{ font-family: Arial, sans-serif; margin: 50px; }}
                h1 {{ color: #e74c3c; }}
            </style>
        </head>
        <body>
            <h1>500 - Server Error</h1>
            <p>An error occurred while processing your request.</p>
            <p>Error: {error}</p>
        </body>
        </html>
        '''
        self.wfile.write(html.encode())

    def log_message(self, format, *args):
        # Custom log format
        print(f"[{self.log_date_time_string()}] {format % args}")


def run_server(port=8000):
    """Run the clean URL development server"""

    # Change to the website directory
    web_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(web_dir)

    # Create server
    with socketserver.TCPServer(("", port), CleanURLHandler) as httpd:
        print("NgKore Website - Clean URL Server")
        print(f"Serving from: {web_dir}")
        print(f"Server URL: http://localhost:{port}")
        print("Clean URLs enabled (no .html extensions in browser)")
        print("\n Test these clean URLs:")
        print(f"   http://localhost:{port}/")
        print(f"   http://localhost:{port}/pages/about")
        print(f"   http://localhost:{port}/pages/contact")
        print(f"   http://localhost:{port}/pages/expertise")
        print(
            f"   http://localhost:{port}/pages/expertise/telecom-architecture")
        print("\n  Press Ctrl+C to stop the server\n")

        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n🛑 Server stopped")


if __name__ == "__main__":
    port = 8000
    if len(sys.argv) > 1:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print("Invalid port number. Using default port 8000.")

    run_server(port)
